src/contracts/test/GPv2SettlementTestInterface.sol

```diff
@@ -18,4 +20,44 @@ contract GPv2SettlementTestInterface is GPv2Settlement {
     function allowanceManagerTest() external view returns (address) {
         return address(allowanceManager);
     }
+
+    function processTradesTest(
+        IERC20[] calldata tokens,
+        uint256[] calldata clearingPrices,
+        bytes calldata encodedTrades
+    )
+        external
+        view
+        returns (
+            GPv2AllowanceManager.Transfer[] memory inTransfers,
+            GPv2AllowanceManager.Transfer[] memory outTransfers
+        )
+    {
+        (inTransfers, outTransfers) = processTrades(
+            tokens,
+            clearingPrices,
+            encodedTrades
+        );
+    }
+
+    function processTradeMemoryTest() external pure returns (uint256 mem) {
+        GPv2Encoding.Trade memory trade;
+        GPv2AllowanceManager.Transfer memory inTransfer;
+        GPv2AllowanceManager.Transfer memory outTransfer;
+
+        // NOTE: Solidity stores the free memory pointer at address 0x40. Read
+        // it before and after calling `processOrder` to ensure that there are
+        // no memory allocations.
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            mem := mload(0x40)
+        }
+
+        processTrade(trade, 1, 1, inTransfer, outTransfer);
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            mem := sub(mload(0x40), mem)
```

Comment:
Excuse my ignorance, but I don't understand the purpose of having all these mloads and other assembly bits all over the place.

Reply:
This is the "free memory pointer": https://docs.soliditylang.org/en/v0.7.5/internals/layout_in_memory.html

Its what Solidity uses to internally keep track of memory.

Reply:
And why are we always freeing it? I can only guess that we are done with whatever memory because it may have been loaded elsewhere (like say after decoding the order bytes from the call data) and we are cleaning up. Does freeing up the memory issue the gas refund at the time the memory is freed allowing us to potentially do more in our method?

Reply:
We aren't freeing it here, just reading it's value to check that it didn't change. If it changed then that would mean that we allocated somewhere that we shouldn't have.

Reply:
Ok thats fine, my question (although placed right here) was meant to be generally asking "what is up with all these scattered tidbits of assembly everywhere throughout all the contracts". At some point, if we have time I would like to go over the reasoning behind each of them at some point...

Reply:
At least now I get @fedgiac's comment

> At this point, I imagine the Solidity compiler fearing Nick's watchful eye every time it dares allocating memory. 😁

Reply:
There shouldn't be any scattered assembly outside of the test interfaces...